### PR TITLE
[Python] Use same contextvars.Context for interceptors and RPC handler

### LINKED ISSUE #33071: [Python Sync] Implement a mechanism to propagate information between interceptors and server handlers.
We recently discovered a bug where interceptors and server handlers were running in different threads. This made it difficult to pass information between them, we need fix it and also and an example for how to propagate information between interceptors and server handlers.

Related issue: https://github.com/grpc/grpc/issues/24470

#### Note
Before we fix it, as a workaround, we recommend pass data using serialized [metadata](https://github.com/grpc/grpc/blob/v1.56.0/src/python/grpcio/grpc/__init__.py#L415).

### PR DESCRIPTION
**Pull request by kcooney**
Fixes #33071

### CODE REVIEW

**Comment by kcooney** [comment]
@XuanWang-Amos would you be a good person to review this?

**Comment by XuanWang-Amos** [comment]
Thanks for the PR!

One thing I want to point out is that the changes you have made are in the Sync stack, while the testing changes are in the Async stack. you'll need change [this file](https://github.com/grpc/grpc/blob/master/src/python/grpcio_tests/tests/unit/_interceptor_test.py) to add the test.

**Comment by kcooney** [comment]
> Thanks for the PR!
> 
> One thing I want to point out is that the changes you have made are in the Sync stack, while the testing changes are in the Async stack. you'll need change [this file](https://github.com/grpc/grpc/blob/master/src/python/grpcio_tests/tests/unit/_interceptor_test.py) to add the test.

Thanks for letting me know. I can hopefully add tests to the correct location in the next week.

What would you like for me to do with the test changes in this PR? They should insure that context locals set in handlers in the sync stack are available to the handler, so I'm guessing these tests would be helpful (but perhaps belong in a different PR).

**Comment by kcooney** [comment]
> One thing I want to point out is that the changes you have made are in the Sync stack, while the testing changes are in the Async stack. you'll need change [this file](https://github.com/grpc/grpc/blob/master/src/python/grpcio_tests/tests/unit/_interceptor_test.py) to add the test.

Done.

**Comment by kcooney** [comment]
@XuanWang-Amos I figured out how to apply the formatter. If someone could update [the wording around here](https://github.com/grpc/grpc/blob/master/CONTRIBUTING.md#generated-project-files) to discuss how to fix formatting that would be fantastic.

In any case, if you could add the `kokoro:run` label again that would be fantastic.

**Comment by XuanWang-Amos** [comment]
Thanks again for your contribution! I'll start another test run.

**Comment by kcooney** [comment]
@XuanWang-Amos looks like one failure and it appears it may be unrelated to my changes.

Amy thoughts as to whether I should spit out the tests for the async stack to a different PR?

**Comment by XuanWang-Amos** [comment]
Sorry about the delay, the failures are fine since they're not related to Python.

I pulled this PR locally and confirmed that the tests will fail if we revert the change. I also verified that the tests are not flaky (failed 0 time out of 1000 bazel test runs for both Sync and Async).

This PR looks good to me, thanks again for working on this!

@gnossen will take a look before we merge it.

**Review by XuanWang-Amos — APPROVED** [review]

**Comment by kcooney** [comment]
Friendly ping @gnossen @XuanWang-Amos

File: src/python/grpcio/grpc/_server.py
Comment (gnossen) [inline src/python/grpcio/grpc/_server.py:125]: Can we add some text to [the public documentation](https://github.com/grpc/grpc/blob/f092b7a73f5d097c411f7cd17d47dab2a839e090/src/python/grpcio/grpc/__init__.py#L1391) about this new guarantee that contextvars will be propagated between server interceptors? Likewise for the asyncio modifications?

**Review by gnossen — APPROVED** [review]
Amazing! This is the most complex contribution we've received from a non-maintainer in quite a while. Very well done. Just one minor note about adding some text to the documentation so people know this guarantee now exists.

File: src/python/grpcio/grpc/_server.py
Comment (kcooney) [reply src/python/grpcio/grpc/_server.py:125]: Done.

Please review the changes. Note did not change the asyncio code, so I'm not 100% sure my description there is correct. Specifically, I'm not sure if/how the code ensures that each request gets a fresh/independent ContextVar instance. The tests try to ensure this, but they don't do this perfectly. This part is of particular interest to me, as I want code inside the handler to be able to set request-local state without that state leaking into future handlers.

If there are changes needed to the wording, would it be possible for you to make them as suggested changes in GitHub and commit them to the branch?

**Comment by gillmylady** [comment]
Thanks for contributing this @kcooney 
Hey @gnossen Do you mind remind us when we are going to release a newer version of grpcio with this change? I'd like to try it too. Previously we only use contextvars for asyncio but not sync. Want to try this out too.

**Comment by jens-getpassport** [comment]
@gnossen any update on this? Was this shipped, and if so, which version?

**Comment by kcooney** [comment]
> Thanks for contributing this @kcooney Hey @gnossen Do you mind remind us when we are going to release a newer version of grpcio with this change? I'd like to try it too. Previously we only use contextvars for asyncio but not sync. Want to try this out too.

@gillmylady @jens-getpassport I'm not a maintainer, but I think this went out with v1.60.0 (see https://github.com/grpc/grpc/compare/v1.59.3...v1.60.0). For some reason this PR was tagged with "release notes: no" so that may be why you don't see it listed in [the release notes for v1.60.0](https://github.com/grpc/grpc/releases/tag/v1.60.0).

**Pull request merged**
